learning/prompts/documentation-lookup/36_ai_kfp_orchestration.md

```diff
@@ -0,0 +1,49 @@
+Prompt:
+How can I orchestrate Apache Beam pipelines with KubeFlow?
+Response:
+Apache Beam is an open source, unified programming model for defining both batch and streaming data-parallel processing pipelines.Each Apache Beam pipeline is a Directed Acyclic Graph (DAG) of data processing operations constructed using Apache Beam SDK in one of the [supported languages](https://beam.apache.org/documentation/sdks/java/).
+To run a pipeline you need to deploy it to one of the supported [Beam runners](https://beam.apache.org/documentation/runners/capability-matrix/).
+You can use Apache Beam as an integral part of a machine learning project for data ingestion, validation, preprocessing, model validation, and inference.
+A full ML workflow also contains other steps such as data exploration, feature engineering, model training, and model deployment. Furthermore, ML workflow needs to track metadata and artifacts for reproducibility and auditability. Popular tools for building and managing ML workflows include [KubeFlow](https://www.kubeflow.org/), [TFX](https://www.tensorflow.org/tfx), and [Apache Airflow](https://airflow.apache.org/).
+
+ `KubeFlow` is an open source ML platform that is designed to enable MLOps - a set of best practices and guiding principles aimed to streamline development and maintenance of ML systems. KubeFlow provides a set of tools for building, deploying, and managing end-to-end ML pipelines in the form of a `DAG` responsible for scheduling and running the pipeline steps and passing execution parameters, metadata, and artifacts between the steps.
+
+You can make Apache Beam pipeline part of a Kubeflow pipeline. In this case Apache Beam pipeline `DAG` becomes a node in the Kubeflow pipeline `DAG`.
+
+In order to run an Apache Beam pipeline as part of a Kubeflow pipeline you need to:
+1. Create KFP components by specifying interfaces and containerizing the component implementation.
+2. Create KFP pipeline by connecting components, and defining how inputs and outputs are passed between the components.
+3. Compile and run KFP pipeline by submitting it to a KFP client endpoint.
+
+Following is an example of a KFP pipeline that orchestrates an Apache Beam preprocessing pipeline:
+```
+    kfp
+    ├── pipeline.py
+    ├── components
+    │   └── preprocess
+    │       ├── Dockerfile
+    │       ├── component.yaml
+    │       ├── requirements.txt
+    │       └── src
+    │           └── preprocess.py
+    └── requirements.txt
+```
+
+Apache Beam pipelines can be used as `KFP` components consisting of a yaml specification `component.yaml` and a python source file `preprocess.py`. The yaml file specifies the input and output arguments of the component, and the python file contains the Apache Beam pipeline code.
+Because `KFP` provides the input and output arguments as command-line arguments, an `argumentparser` is needed. Component logic and requirements are containerized and the container image is pushed to a container registry.
```

Comment:
```suggestion
Since KFP provides input and output arguments as command-line arguments, an `ArgumentParser` is required. The component logic and requirements are containerized, and the container image is pushed to a container registry.
```